modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -416,17 +416,44 @@ private void onPlacementDriverMessageReceived(NetworkMessage msg0, ClusterNode s
         }
 
         try {
-            CompletableFuture<Replica> replicaFut = replicas.computeIfAbsent(msg.groupId(), k -> new CompletableFuture<>());
-
-            replicaFut
-                    .thenCompose(replica -> replica.processPlacementDriverMessage(msg))
-                    .whenComplete((response, ex) -> {
-                        if (ex == null) {
-                            clusterNetSvc.messagingService().respond(senderConsistentId, response, correlationId);
-                        } else if (!(unwrapCause(ex) instanceof NodeStoppingException)) {
-                            LOG.error("Failed to process placement driver message [msg={}].", ex, msg);
-                        }
-                    });
+            Set<ReplicationGroupId> replicationGroupIds = zonePartIdToTablePartId.get((ZonePartitionId) msg.groupId());
+
+            CompletableFuture<LeaseGrantedMessageResponse>[] futures = new CompletableFuture[replicationGroupIds.size()];
+
+            int i = 0;
+
+            for (ReplicationGroupId grpId : replicationGroupIds) {
+                CompletableFuture<Replica> replicaFut = replicas.computeIfAbsent(grpId, k -> new CompletableFuture<>());
+                futures[i++] = replicaFut.thenCompose(replica -> replica.processPlacementDriverMessage(msg));
+            }
```

Comment:
Missed in the past, but the way to loop this looks strange. The suggestion below may be a little less performant, but more readable, I may guess.

```suggestion
CompletableFuture<LeaseGrantedMessageResponse>[] futures = 
                    (CompletableFuture<LeaseGrantedMessageResponse>[]) replicationGroupIds
                    .stream()
                    .map(grpId -> replicas
                            .computeIfAbsent(grpId, k -> new CompletableFuture<>())
                            .thenCompose(replica -> replica.processPlacementDriverMessage(msg)))
                    .toArray();
```

Reply:
Here we are at the point where the simple loop is less clear than the streaming API...
I am not sure that the code needs to be changed, but if anyone else thinks that, I am ready to apply the proposal.

Reply:
>Here we are at the point where the simple loop is less clear than the streaming API

It's fine, at least I understood it, but `int i` in the middle of a long method + `[i++]` indexing looking worse than it could be imo.